File: cli/node/src/option-builder.ts
/**
 * Reusable CLI option builders derived from the shared CLI contract.
 *
 * This module centralizes option definitions, validation rules, and default
 * values so that every command uses consistent parameter names, help text,
 * and defaults.
 *
 * All option builders follow the contract defined in `cli/cli-spec.json`
 * and `src/contract/payload_contract.json`.
 */

import { Command, Option } from 'commander';
import { VALIDATION_RULES, validateField } from './schema/fields.js';
import contract from './contract/payload_contract.json' with { type: 'json' };

interface ContractDefaults {
  top_k?: number;
  threshold?: number;
  rerank?: boolean;
  keyword?: boolean;
  immutable?: boolean;
  [key: string]: unknown;
}

const defaults = (contract.defaults || {}) as ContractDefaults;

export interface OptionSpec {
  flags: string;
  description: string;
  defaultValue?: unknown;
  choices?: string[];
  required?: boolean;
  argParser?: (value: string) => unknown;
  helpGroup?: string;
}

export type OptionGroup = Record<string, OptionSpec>;

export function scopeOptions(helpGroup = 'Scope'): OptionGroup {
  return {
    userId: {
      flags: '--user-id <id>',
      description: 'Scope to user.',
      helpGroup,
    },
    agentId: {
      flags: '--agent-id <id>',
      description: 'Scope to agent.',
      helpGroup,
    },
    appId: {
      flags: '--app-id <id>',
      description: 'Scope to app.',
      helpGroup,
    },
    runId: {
      flags: '--run-id <id>',
      description: 'Scope to run.',
      helpGroup,
    },
  };
}

export function outputOption(
  defaultFormat = 'text',
  choices = ['text', 'json', 'table', 'quiet'],
  helpGroup = 'Output',
): OptionSpec {
  const choicesStr = choices.join(', ');
  return {
    flags: '--output <format>',
    description: `Output format: ${choicesStr}.`,
    defaultValue: defaultFormat,
    choices,
    helpGroup,
  };
}

export function connectionOptions(helpGroup = 'Connection'): OptionGroup {
  return {
    apiKey: {
      flags: '--api-key <key>',
      description: 'Override API key.',
      helpGroup,
    },
    baseUrl: {
      flags: '--base-url <url>',
      description: 'Override API base URL.',
      helpGroup,
    },
  };
}

export function searchOptions(helpGroup = 'Search'): OptionGroup {
  return {
    topK: {
      flags: '--top-k <n>',
      description: 'Number of results.',
      defaultValue: defaults.top_k ?? 10,
      argParser: parseInt,
      helpGroup,
    },
    threshold: {
      flags: '--threshold <n>',
      description: 'Minimum similarity score.',
      defaultValue: defaults.threshold ?? 0.3,
      argParser: parseFloat,
      helpGroup,
    },
    rerank: {
      flags: '--rerank',
      description: 'Enable reranking (Platform only).',
      defaultValue: defaults.rerank ?? false,
      helpGroup,
    },
    keyword: {
      flags: '--keyword',
      description: 'Use keyword search.',
      defaultValue: defaults.keyword ?? false,
      helpGroup,
    },
    filter: {
      flags: '--filter <json>',
      description: 'Advanced filter expression (JSON).',
      helpGroup,
    },
    fields: {
      flags: '--fields <fields>',
      description: 'Specific fields to return (comma-separated).',
      helpGroup,
    },
  };
}

export function paginationOptions(helpGroup = 'Pagination'): OptionGroup {
  return {
    page: {
      flags: '--page <n>',
      description: 'Page number.',
      defaultValue: 1,
      argParser: parseInt,
      helpGroup,
    },
    pageSize: {
      flags: '--page-size <n>',
      description: 'Results per page.',
      defaultValue: 100,
      argParser: parseInt,
      helpGroup,
    },
  };
}

export function filterOptions(helpGroup = 'Filters'): OptionGroup {
  return {
    category: {
      flags: '--category <name>',
      description: 'Filter by category.',
      helpGroup,
    },
    after: {
      flags: '--after <date>',
      description: 'Created after (YYYY-MM-DD).',
      helpGroup,
    },
    before: {
      flags: '--before <date>',
      description: 'Created before (YYYY-MM-DD).',
      helpGroup,
    },
  };
}

export function addOptions(): OptionGroup {
  return {
    messages: {
      flags: '--messages <json>',
      description: 'Conversation messages as JSON.',
    },
    file: {
      flags: '-f, --file <path>',
      description: 'Read messages from JSON file.',
    },
    metadata: {
      flags: '-m, --metadata <json>',
      description: 'Custom metadata as JSON.',
    },
    immutable: {
      flags: '--immutable',
      description: 'Prevent future updates.',
      defaultValue: defaults.immutable ?? false,
    },
    noInfer: {
      flags: '--no-infer',
      description: 'Skip inference, store raw.',
      defaultValue: false,
    },
    expires: {
      flags: '--expires <date>',
      description: 'Expiration date (YYYY-MM-DD).',
    },
    categories: {
      flags: '--categories <list>',
      description: 'Categories (JSON array or comma-separated).',
    },
  };
}

export function applyOptions(cmd: Command, options: OptionGroup): Command {
  for (const [_name, spec] of Object.entries(options)) {
    const opt = new Option(spec.flags, spec.description);

    if (spec.defaultValue !== undefined) {
      opt.default(spec.defaultValue);
    }
    if (spec.choices?.length) {
      opt.choices(spec.choices);
    }
    const optAny = opt as unknown as Record<string, unknown>;
    if (spec.helpGroup && typeof optAny.helpGroup === 'function') {
      (optAny.helpGroup as (name: string) => Option)(spec.helpGroup);
    }
    if (spec.argParser) {
      opt.argParser(spec.argParser as (value: string, previous: unknown) => unknown);
    }

    cmd.addOption(opt);
  }
  return cmd;
}

export function validateTopK(topK: number): void {
  validateField('top_k', topK);
}

export function validateThreshold(threshold: number): void {
  validateField('threshold', threshold);
}

export function validatePage(page: number): void {
  validateField('page', page);
}

export function validatePageSize(pageSize: number): void {
  validateField('pageSize', pageSize);
}

export function parseJsonOption(value: string | undefined, name: string): unknown {
  if (value === undefined || value === null) return null;
  try {
    return JSON.parse(value);
  } catch (e) {
    throw new Error(`Invalid JSON in ${name}: ${e instanceof Error ? e.message : String(e)}`);
  }
}

export function parseCategories(value: string | undefined): string[] | null {
  if (value === undefined || value === null) return null;
  try {
    const parsed = JSON.parse(value);
    if (!Array.isArray(parsed)) {
      throw new Error('--categories JSON must be an array.');
    }
    return parsed.map(String);
  } catch {
    return value.split(',').map((c) => c.trim()).filter(Boolean);
  }
}

export function parseFields(value: string | undefined): string[] | null {
  if (value === undefined || value === null) return null;
  return value.split(',').map((f) => f.trim()).filter(Boolean);
}

export function buildScope(params: {
  userId?: string;
  agentId?: string;
  appId?: string;
  runId?: string;
}): Record<string, string> | null {
  const scope: Record<string, string> = {};
  if (params.userId) scope.user_id = params.userId;
  if (params.agentId) scope.agent_id = params.agentId;
  if (params.appId) scope.app_id = params.appId;
  if (params.runId) scope.run_id = params.runId;
  return Object.keys(scope).length ? scope : null;
}

File: cli/python/src/mem0_cli/option_builder.py
"""Reusable CLI option builders derived from the shared CLI contract.

This module centralizes option definitions, validation rules, and default values
so that every command uses consistent parameter names, help text, and defaults.

All option builders follow the contract defined in ``cli/cli-spec.json``.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from mem0.schema.fields import FIELD_VALIDATION

CONTRACT_PATH = Path(__file__).parent / "contract" / "payload_contract.json"


def _load_contract() -> dict[str, Any]:
    with open(CONTRACT_PATH) as f:
        return json.load(f)


_contract = _load_contract()
_defaults: dict[str, Any] = _contract.get("defaults", {})


def scope_options(
    *,
    help_panel: str = "Scope",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard scope option group.

    Returns a dict of keyword arguments suitable for ``typer.Option()`` calls.
    Each key is the snake_case option name; each value is a dict of
    ``{"default": ..., "param_decls": [...], "help": ..., "rich_help_panel": ...}``.
    """
    return {
        "user_id": {
            "default": None,
            "param_decls": ["--user-id", "-u"],
            "help": "Scope to user.",
            "rich_help_panel": help_panel,
        },
        "agent_id": {
            "default": None,
            "param_decls": ["--agent-id"],
            "help": "Scope to agent.",
            "rich_help_panel": help_panel,
        },
        "app_id": {
            "default": None,
            "param_decls": ["--app-id"],
            "help": "Scope to app.",
            "rich_help_panel": help_panel,
        },
        "run_id": {
            "default": None,
            "param_decls": ["--run-id"],
            "help": "Scope to run.",
            "rich_help_panel": help_panel,
        },
    }


def output_option(
    *,
    default: str = "text",
    choices: list[str] | None = None,
    help_panel: str = "Output",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard --output option."""
    choices_str = ", ".join(choices) if choices else "text, json, table, quiet"
    return {
        "default": default,
        "param_decls": ["--output", "-o"],
        "help": f"Output format: {choices_str}.",
        "rich_help_panel": help_panel,
    }


def connection_options(
    *,
    help_panel: str = "Connection",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard connection option group."""
    return {
        "api_key": {
            "default": None,
            "param_decls": ["--api-key"],
            "help": "Override API key.",
            "rich_help_panel": help_panel,
            "envvar": "MEM0_API_KEY",
        },
        "base_url": {
            "default": None,
            "param_decls": ["--base-url"],
            "help": "Override API base URL.",
            "rich_help_panel": help_panel,
        },
    }


def search_options(
    *,
    help_panel: str = "Search",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard search option group."""
    return {
        "top_k": {
            "default": _defaults.get("top_k", 10),
            "param_decls": ["--top-k", "-k", "--limit"],
            "help": "Number of results.",
            "rich_help_panel": help_panel,
        },
        "threshold": {
            "default": _defaults.get("threshold", 0.3),
            "param_decls": ["--threshold"],
            "help": "Minimum similarity score.",
            "rich_help_panel": help_panel,
        },
        "rerank": {
            "default": _defaults.get("rerank", False),
            "param_decls": ["--rerank"],
            "help": "Enable reranking (Platform only).",
            "rich_help_panel": help_panel,
        },
        "keyword": {
            "default": _defaults.get("keyword", False),
            "param_decls": ["--keyword"],
            "help": "Use keyword search.",
            "rich_help_panel": help_panel,
        },
        "filter_json": {
            "default": None,
            "param_decls": ["--filter"],
            "help": "Advanced filter expression (JSON).",
            "rich_help_panel": help_panel,
        },
        "fields": {
            "default": None,
            "param_decls": ["--fields"],
            "help": "Specific fields to return (comma-separated).",
            "rich_help_panel": help_panel,
        },
    }


def pagination_options(
    *,
    help_panel: str = "Pagination",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard pagination option group."""
    return {
        "page": {
            "default": 1,
            "param_decls": ["--page"],
            "help": "Page number.",
            "rich_help_panel": help_panel,
        },
        "page_size": {
            "default": 100,
            "param_decls": ["--page-size"],
            "help": "Results per page.",
            "rich_help_panel": help_panel,
        },
    }


def filter_options(
    *,
    help_panel: str = "Filters",
) -> dict[str, Any]:
    """Return Typer option kwargs for the standard filter option group."""
    return {
        "category": {
            "default": None,
            "param_decls": ["--category"],
            "help": "Filter by category.",
            "rich_help_panel": help_panel,
        },
        "after": {
            "default": None,
            "param_decls": ["--after"],
            "help": "Created after (YYYY-MM-DD).",
            "rich_help_panel": help_panel,
        },
        "before": {
            "default": None,
            "param_decls": ["--before"],
            "help": "Created before (YYYY-MM-DD).",
            "rich_help_panel": help_panel,
        },
    }


def add_options() -> dict[str, Any]:
    """Return Typer option kwargs for the standard add command options."""
    return {
        "messages": {
            "default": None,
            "param_decls": ["--messages"],
            "help": "Conversation messages as JSON.",
        },
        "file": {
            "default": None,
            "param_decls": ["--file", "-f"],
            "help": "Read messages from JSON file.",
        },
        "metadata": {
            "default": None,
            "param_decls": ["--metadata", "-m"],
            "help": "Custom metadata as JSON.",
        },
        "immutable": {
            "default": _defaults.get("immutable", False),
            "param_decls": ["--immutable"],
            "help": "Prevent future updates.",
        },
        "no_infer": {
            "default": False,
            "param_decls": ["--no-infer"],
            "help": "Skip inference, store raw.",
        },
        "expires": {
            "default": None,
            "param_decls": ["--expires"],
            "help": "Expiration date (YYYY-MM-DD).",
        },
        "categories": {
            "default": None,
            "param_decls": ["--categories"],
            "help": "Categories (JSON array or comma-separated).",
        },
    }


def validate_top_k(top_k: int) -> None:
    """Validate top_k against the contract rules."""
    if top_k < 1:
        raise ValueError(FIELD_VALIDATION["top_k"]["error"])


def validate_threshold(threshold: float) -> None:
    """Validate threshold against the contract rules."""
    if not (0.0 <= threshold <= 1.0):
        raise ValueError(FIELD_VALIDATION["threshold"]["error"])


def validate_page(page: int) -> None:
    """Validate page number against the contract rules."""
    if page < 1:
        raise ValueError(FIELD_VALIDATION["page"]["error"])


def validate_page_size(page_size: int) -> None:
    """Validate page size against the contract rules."""
    if page_size < 1:
        raise ValueError(FIELD_VALIDATION["page_size"]["error"])


def parse_json_option(value: str | None, *, name: str) -> Any:
    """Parse a JSON-encoded CLI option, raising a ValueError on failure.

    Parameters
    ----------
    value:
        Raw string from the CLI, or ``None`` if not provided.
    name:
        Option name used in error messages (e.g. ``"--filter"``).
    """
    if value is None:
        return None
    try:
        return json.loads(value)
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON in {name}: {e}") from None


def parse_categories(value: str | None) -> list[str] | None:
    """Parse --categories: try JSON array first, fall back to comma-split."""
    if value is None:
        return None
    try:
        parsed = json.loads(value)
        if not isinstance(parsed, list):
            raise ValueError("--categories JSON must be an array.")
        return [str(x) for x in parsed]
    except json.JSONDecodeError:
        return [c.strip() for c in value.split(",")]


def parse_fields(value: str | None) -> list[str] | None:
    """Parse --fields: comma-separated list."""
    if value is None:
        return None
    return [f.strip() for f in value.split(",")]


def build_scope(
    *,
    user_id: str | None = None,
    agent_id: str | None = None,
    app_id: str | None = None,
    run_id: str | None = None,
) -> dict[str, str] | None:
    """Build a scope dict from entity IDs, returning None if all are empty.

    The scope dict only includes keys with non-None values, matching the
    JSON envelope contract.
    """
    scope: dict[str, str] = {}
    if user_id:
        scope["user_id"] = user_id
    if agent_id:
        scope["agent_id"] = agent_id
    if app_id:
        scope["app_id"] = app_id
    if run_id:
        scope["run_id"] = run_id
    return scope or None
